server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java

```diff
@@ -0,0 +1,346 @@
+  private final List<String> parentsIBuilt = new CopyOnWriteArrayList<String>();
+
+  private boolean started = false;
+
+  public NodeAnnouncer(CuratorFramework curator)
+  {
+    this.curator = curator;
+  }
+
+  @VisibleForTesting
+  Set<String> getAddedPaths()
+  {
+    return announcements.keySet();
+  }
+
+  @LifecycleStart
+  public void start()
+  {
+    log.info("Starting announcer");
+    synchronized (toAnnounce) {
+      if (started) {
+        return;
+      }
+
+      started = true;
+
+      for (Announceable announceable : toAnnounce) {
+        announce(announceable.path, announceable.bytes, announceable.removeParentsIfCreated);
+      }
+      toAnnounce.clear();
+
+      for (Announceable announceable : toUpdate) {
+        update(announceable.path, announceable.bytes);
+      }
+      toUpdate.clear();
+    }
+  }
+
+  @LifecycleStop
+  public void stop()
+  {
+    log.info("Stopping announcer");
+    synchronized (toAnnounce) {
+      if (!started) {
+        return;
+      }
+
+      started = false;
+
+      Closer closer = Closer.create();
+      for (NodeCache cache : listeners.values()) {
+        closer.register(cache);
+      }
+      CloseQuietly.close(closer);
+
+      for (String announcementPath : announcements.keySet()) {
+        unannounce(announcementPath);
+      }
+
+      if (!parentsIBuilt.isEmpty()) {
+        CuratorTransaction transaction = curator.inTransaction();
+        for (String parent : parentsIBuilt) {
+          try {
+            transaction = transaction.delete().forPath(parent).and();
+          }
+          catch (Exception e) {
+            log.info(e, "Unable to delete parent[%s], boooo.", parent);
+          }
+        }
+        try {
+          ((CuratorTransactionFinal) transaction).commit();
+        }
+        catch (Exception e) {
+          log.info(e, "Unable to commit transaction. Please feed the hamsters");
+        }
+      }
+    }
+  }
+
+  /**
+   * Like announce(path, bytes, true).
+   */
+  public void announce(String path, byte[] bytes)
+  {
+    announce(path, bytes, true);
+  }
+
+  /**
+   * Announces the provided bytes at the given path.  Announcement means that it will create an ephemeral node
+   * and monitor it to make sure that it always exists until it is unannounced or this object is closed.
+   *
+   * @param path                  The path to announce at
+   * @param bytes                 The payload to announce
+   * @param removeParentIfCreated remove parent of "path" if we had created that parent
+   */
+  public void announce(String path, byte[] bytes, boolean removeParentIfCreated)
+  {
+    synchronized (toAnnounce) {
+      if (!started) {
+        toAnnounce.add(new Announceable(path, bytes, removeParentIfCreated));
+        return;
+      }
+    }
+
+    final ZKPaths.PathAndNode pathAndNode = ZKPaths.getPathAndNode(path);
+
+    final String parentPath = pathAndNode.getPath();
```

Comment:
Please extract it to a static utility method `ZKPathsUtils.getParentPath(path) { return ZKPaths.getPathAndNode(path).getPath(); }`, because currently the code looks confusing.